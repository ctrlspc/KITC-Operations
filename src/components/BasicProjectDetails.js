import React from 'react';

export default ({title, description, projectManager, projectType}) => (
  <div>
    <h3>{title}</h3>
    <p>{description}</p>
    <p><span>Project Manager:</span> {projectManager.name}</p>
  </div>
);