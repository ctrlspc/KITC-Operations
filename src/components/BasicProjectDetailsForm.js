import React from 'react';
import _ from 'lodash';
export default class BasicProjectDetailsForm extends React.Component {
 
  constructor(props) {
    super(props);
    const project = this.props.project;
    this.state = {
      title: project ? project.title : '',
      description: project ? project.description : '',
      projectManager: project ? project.projectManager.uid  : '' ,
      projectType: project ? project.projectType : 'ext',
      error:''
    };
  };

  componentDidUpdate(prevProps, prevState, snapshot) {
    if (prevProps.projectManagers !== this.props.projectManagers) {
      if (this.state.projectManager === '' && !_.isEmpty(this.props.projectManagers[0])) {
        this.setState(() => ({projectManager:this.props.projectManagers[0].uid}));
      }
    }
  };

  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({title}));
  };

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));
  };

  onManagerChange = (e) => {
    const projectManager = e.target.value;
    this.setState(() => ({projectManager}));
  };

  onTypeChange = (e) => {
    const projectType = e.target.value;
    this.setState(() => ({projectType}));
  };

  onSubmit = (e) => {
    e.preventDefault();

    const projectManager = this.props.projectManagers.find((projectManager) => {
      return projectManager.uid === this.state.projectManager
    });
    if(!this.state.title || !this.state.description || !this.state.projectManager) {
      this.setState(() => ({error:'Please provide a title and project description'}));
    } else {

      this.setState(() => ({error:''}));
      this.props.onSubmit({
        title: this.state.title,
        description: this.state.description,
        projectType: this.state.projectType,
        projectManager
      });
    }
  };

  render () {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p> }
        <input 
          autoFocus
          className="text-input"
          type="text" 
          placeholder="Project Name"
          value={this.state.title}
          onChange={this.onTitleChange}
        />
        <textarea 
          className="textarea"
          placeholder="Project Description"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <div className="select-group">
          <div className="select-group__item">
            <h3 className="select-group__title">Project Manager</h3>
            <select 
              className="select-group__select"
              id="project-manager"
              onChange={this.onManagerChange}
              value={this.state.projectManager}
            >
              {this.props.projectManagers && this.props.projectManagers.map((item) => {
                return <option key={item.uid} value={item.uid}>{item.displayName}</option> 
              })}
            </select>
          </div>
          <div className="select-group__item">
            <h3 className="select-group__title">Project Type</h3>
            <select 
              className="select-group__select"
              id="project-type" 
              onChange={this.onTypeChange}
              value={this.state.projectType}
            >
              <option value="int">Internal</option>
              <option value="ext">External</option>
            </select>
          </div>
        </div>
        <div>
          <button className="button">Save Project</button>
        </div>
        
      </form>
    );
  };
};
