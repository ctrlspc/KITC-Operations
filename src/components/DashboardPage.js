import React from 'react';
import ProjectList from './ProjectList';
import { Link } from 'react-router-dom';

const DashboardPage = () => (

  <div>
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Projects Dashboard
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Create a Project</Link>
        </div>
      </div>
    </div>
    <div className="content-container">
      <ProjectList projects={[
        {
          id:'123',
          title:'A project',
          description:'a description',
          projectManager:{id:'1',name:'Jason Marshall'},
          projectType:'int'
        }
      ]} />
    </div>
  </div>
  
);

export default DashboardPage;