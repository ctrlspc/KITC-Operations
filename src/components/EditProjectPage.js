import React from 'react';
import { connect } from 'react-redux';
import ProjectForm from '../components/ProjectForm';
import { updateProject } from '../actions/projects';

export class EditProjectPage extends React.Component {

  onSubmit = (project) => {
    this.props.updateProject(this.props.project.id, project);
    this.props.history.push('/');
  }

  render () {
    return (
      <div className="content-container">
        <h1>Edit an existing Project</h1>
        <ProjectForm 
          onSubmit={this.onSubmit} 
          projectManagers={this.props.projectManagers}
          project={this.props.project}
        />
      </div>
    );
  };
};

export const mapStateToProps = (state, props) => ({
  project: state.projects.find((project) => project.id === props.match.params.id),
  projectManagers: state.projectManagers
})

export const mapDispatchToProps = (dispatch) => ({
  updateProject: (id, project) => dispatch(updateProject(id, project))
})


export default connect(mapStateToProps,mapDispatchToProps)(EditProjectPage);