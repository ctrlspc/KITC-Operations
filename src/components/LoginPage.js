import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/iam';

export const LoginPage = ({ startLogin }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">KITC</h1>
      <p>Project Operations Tracker</p>
      <button className="button" onClick={startLogin}>Login with Google</button>
    </div>
    
  </div>
);

export const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);