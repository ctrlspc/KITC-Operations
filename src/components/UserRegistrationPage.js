import React from 'react';
import { connect } from 'react-redux';
import { startRegisterUser } from '../actions/iam';
import _ from "lodash";

export class UserRegistrationPage extends React.Component {

  onRegisterButtonClick = (e) => {
    e.preventDefault();
    this.props.onRegister(this.props.user);
  };
  
  render () {
    return (
      <div className="box-layout">
        <div className="box-layout__box">
          <h1 className="box-layout__title">KITC Operations Registration</h1>
          { _.isEmpty(this.props.profile) ? (
            <div>
              <p>You can only use this app if you work for the KITC. If you do, great, hit register below!</p>
              <button className="button" onClick={this.onRegisterButtonClick}>Register</button>
            </div>
          ) : (
            _.isEmpty(this.props.roles) ? 
            (
              <p>Your registration is being reviewed by an administrator. You'll be notified by email when your account is activated.</p>
            ) : (
              <p>You're all set - you can continue over in the <a href="/dashboard">Dashboard</a></p>
            )
          )
          }
        </div>
      </div>
    );
  };
}; 

export const mapStateToProps = (state) => {
  console.log(state.iam);
  
  return {
  user: state.iam.identity,
  profile: state.iam.profile,
  roles: state.iam.roles
  }
};

export const mapDispatchToProps = (dispatch) => ({
  onRegister: (user) => dispatch(startRegisterUser(user))
});

export default connect(mapStateToProps, mapDispatchToProps)(UserRegistrationPage);

