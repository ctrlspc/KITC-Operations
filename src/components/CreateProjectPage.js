import React from 'react';
import { connect } from 'react-redux';

import BasicProjectDetailsForm from './BasicProjectDetailsForm';
import { startAddProject } from '../actions/projects';
import { getActiveUsers } from '../reducers';
import { startSetUsers } from '../actions/users';

export class CreateProjectPage extends React.Component {

  constructor(props) {
    super(props);
    this.props.getUsers();
  }

  onSubmit = (project) => {
    this.props.addProject(project);
    this.props.history.push('/');
  };

  render () {
    return (
      <div className="content-container">
        <h1>Create a new Project</h1>
        <BasicProjectDetailsForm 
          onSubmit={this.onSubmit} 
          projectManagers={this.props.users}
        />
      </div>
    );
  };
};

export const mapDispatchToProps = (dispatch) => ({
  getUsers: () => dispatch(startSetUsers()),
  addProject: (project) => dispatch(startAddProject(project))
});

export const mapStateToProps = (state) => ({
  users: getActiveUsers(state)
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateProjectPage);


