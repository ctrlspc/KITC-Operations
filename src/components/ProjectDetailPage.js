import React from 'react';
import { connect } from 'react-redux';
import BasicProjectDetails from './BasicProjectDetails';
import BasicProjectDetailsForm from './BasicProjectDetailsForm';
import { startUpdateProject } from '../actions/projects';
import { getActiveUsers } from '../reducers';
import { startSetUsers } from '../actions/users';

export class ProjectDetailPage extends React.Component {

  constructor(props) {
    super(props);
    this.props.getUsers();
    this.state = {
      editBasicProjectDetails: !!props.editBasicProjectDetails
    };
  }

  onEditBasicProjectDetails = () => {
    this.setState(() => ({editBasicProjectDetails: true}));
  };

  onFinishedEditingBasicProjectDetails = (updates) => {
    this.props.updateProject(this.props.project.id, updates).then(() => {
      this.setState(() => ({editBasicProjectDetails: false}));
    });
  }

  render (){
    return (
      <div className="content-container">
        <div className="box-section">
          <div className="box-section__header">
            <h1>Basic Project Details</h1>
            {!this.state.editBasicProjectDetails && 
              <button className="button" onClick={this.onEditBasicProjectDetails}>Edit</button>
            }
          </div>
          <div className="box-section__body">
            {
              this.state.editBasicProjectDetails ? 
              (
                <BasicProjectDetailsForm 
                  onSubmit={this.onFinishedEditingBasicProjectDetails} 
                  projectManagers={this.props.users}
                  project={this.props.project}
                />
              ) : (
                <BasicProjectDetails {...this.props.project} onEditClick={this.onEditBasicProjectDetails}/>
              )
            }
          </div>
        </div>
      </div>
    );
  };
};

export const mapStateToProps = (state, props) => ({
  project: state.projects.find((project) => project.id === props.match.params.id),
  users: getActiveUsers(state)
});

export const mapDispatchToProps = (dispatch) => ({
  getUsers: () => dispatch(startSetUsers()),
  updateProject: (id, project) => dispatch(startUpdateProject(id, project))
})

export default connect(mapStateToProps, mapDispatchToProps)(ProjectDetailPage);