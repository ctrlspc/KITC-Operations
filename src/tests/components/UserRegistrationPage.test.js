import React from 'react';
import { shallow } from 'enzyme';
import { UserRegistrationPage, mapStateToProps, mapDispatchToProps } from '../../components/UserRegistrationPage';

test('should render the ProjectListItem component correctly for a user who has not registered', () => {
  const user = {uid:'123'};
  const profile = {};
  const roles = [];
  const wrapper = shallow (<UserRegistrationPage user={user} profile={profile} roles={roles}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render the ProjectListItem component correctly for a user who has not been assigned any roles', () => {
  const user = {uid:'123'};
  const profile = {uid:'123'};
  const roles = [];
  const wrapper = shallow (<UserRegistrationPage user={user} profile={profile} roles={roles}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render the ProjectListItem component correctly for a user who has completed registration', () => {
  const user = {uid:'123'};
  const profile = {uid:'123'};
  const roles = ['user'];
  const wrapper = shallow (<UserRegistrationPage user={user} profile={profile} roles={roles}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should call onRegisterButtonClick on button click', () => {
  const user = {uid:'123'};
  const profile = {};
  const roles = [];
  const onRegister = jest.fn();
  const event = {preventDefault: jest.fn()};
  const wrapper = shallow (<UserRegistrationPage onRegister={onRegister} user={user} profile={profile} roles={roles}/>);

  wrapper.find('button').simulate('click',event);

  expect(onRegister).toHaveBeenCalledWith({uid:'123'});
  expect(event.preventDefault).toHaveBeenCalled();
});

test('should map State to Props correctly', () => {
  const state = {
    iam: {
      identity:{uid:'123'},
      profile:{uid:'123'},
      roles:['user']
    }
  };

  const props = mapStateToProps(state);

  expect(props.user).toEqual(state.iam.identity);
  expect(props.profile).toEqual(state.iam.profile);
  expect(props.roles).toEqual(state.iam.roles);

});