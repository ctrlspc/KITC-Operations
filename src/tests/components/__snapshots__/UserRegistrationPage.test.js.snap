// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the ProjectListItem component correctly for a user who has completed registration 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      KITC Operations Registration
    </h1>
    <div>
      <p>
        You can only use this app if you work for the KITC. If you do, great, hit register below!
      </p>
      <button
        className="button"
        onClick={[Function]}
      >
        Register
      </button>
    </div>
  </div>
</div>
`;

exports[`should render the ProjectListItem component correctly for a user who has not been assigned any roles 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      KITC Operations Registration
    </h1>
    <div>
      <p>
        You can only use this app if you work for the KITC. If you do, great, hit register below!
      </p>
      <button
        className="button"
        onClick={[Function]}
      >
        Register
      </button>
    </div>
  </div>
</div>
`;

exports[`should render the ProjectListItem component correctly for a user who has not registered 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      KITC Operations Registration
    </h1>
    <div>
      <p>
        You can only use this app if you work for the KITC. If you do, great, hit register below!
      </p>
      <button
        className="button"
        onClick={[Function]}
      >
        Register
      </button>
    </div>
  </div>
</div>
`;
